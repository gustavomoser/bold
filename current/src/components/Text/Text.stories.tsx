import { number, select, text } from '@storybook/addon-knobs'
import React from 'react'

import { TextColor } from '../../styles'

import { Text, TextProps } from './Text'

const variants: Array<TextProps['variant']> = [
  'main',
  'secondary',
  'disabled',
  'link',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
]
const components: Array<TextProps['component']> = ['span', 'p', 'button']
const fontWeights: Array<TextProps['fontWeight']> = ['' as any, 'normal', 'bold']
const fontStyles: Array<TextProps['fontStyle']> = ['' as any, 'normal', 'italic']
const textDecorations: Array<TextProps['textDecoration']> = ['' as any, 'normal', 'underline']
const colors: TextColor[] = [
  '' as any,
  'inherit',
  'normal',
  'secondary',
  'disabled',
  'primary',
  'danger',
  'info',
  'alert',
  'success',
]

export default {
  title: 'Components/Textual',
}

export const _Text = () => (
  <Text
    variant={select('variant', variants, undefined)}
    component={select('component', components as any, undefined)}
    fontSize={number('fontSize', undefined)}
    color={select('color', colors, undefined)}
    fontWeight={select('fontWeight', fontWeights, undefined)}
    fontStyle={select('fontStyle', fontStyles, undefined)}
    textDecoration={select('textDecoration', textDecorations, undefined)}
  >
    {text('text', 'Lorem ipsum')}
  </Text>
)
