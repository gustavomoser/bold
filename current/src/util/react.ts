import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
