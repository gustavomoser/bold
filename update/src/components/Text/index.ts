export { Text } from './Text'
export type { BaseTextProps, TextProps } from './Text'
