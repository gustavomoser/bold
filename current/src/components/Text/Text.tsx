import { FontStyleProperty, FontWeightProperty, TextDecorationProperty } from 'csstype'
import React, { CSSProperties, HTMLAttributes } from 'react'

import { ExternalStyles, TextColor, useStyles } from '../../styles'
import { getTextColor, Theme } from '../../styles/theme/createTheme'
import { TypographyVariant } from '../../styles/theme/createTypography'
import { Omit } from '../../util'

export interface BaseTextProps {
  component?: React.ElementType
  variant?: TypographyVariant
  color?: TextColor
  fontSize?: number
  fontWeight?: FontWeightProperty
  fontStyle?: FontStyleProperty
  textDecoration?: TextDecorationProperty<string>
  style?: ExternalStyles
  [key: string]: any
}

export type TextProps = Omit<HTMLAttributes<HTMLElement>, 'style'> & BaseTextProps

export function Text<P = TextProps>(props: P & BaseTextProps) {
  const { component, variant, color, fontSize, fontWeight, fontStyle, textDecoration, style, ...rest } = props
  const { classes, css } = useStyles(createStyles, props)

  const Cmp = component

  const className = css(
    classes.root,
    color && classes.color,
    fontSize && classes.fontSize,
    fontWeight && classes.fontWeight,
    fontStyle && classes.fontStyle,
    textDecoration && classes.textDecoration,
    style
  )

  return <Cmp className={className} {...rest} />
}

Text.defaultProps = {
  component: 'span',
  variant: 'main',
}

const createStyles = (
  theme: Theme,
  { variant, color, fontWeight, fontStyle, textDecoration, fontSize }: BaseTextProps
) => ({
  root: {
    ...theme.typography.variant(variant),
  } as CSSProperties,
  color: {
    color: color && getTextColor(theme, color),
  } as CSSProperties,
  fontSize: {
    fontSize: fontSize && `${fontSize}rem`,
  } as CSSProperties,
  fontWeight: {
    fontWeight,
  } as CSSProperties,
  fontStyle: {
    fontStyle,
  } as CSSProperties,
  textDecoration: {
    textDecoration,
  } as CSSProperties,
})
