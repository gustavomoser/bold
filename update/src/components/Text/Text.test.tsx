import { render } from '@testing-library/react'
import React, { AnchorHTMLAttributes } from 'react'

import { Text } from './Text'

it('should render correctly', () => {
  expect(render(<Text>Test 1</Text>).container).toMatchSnapshot()
})

it('should accept text style decoration props', () => {
  expect(render(<Text fontWeight='bold'>Test weight</Text>).container).toMatchSnapshot()
  expect(render(<Text fontSize={2}>Test size</Text>).container).toMatchSnapshot()
  expect(render(<Text color='primary'>Test color</Text>).container).toMatchSnapshot()
  expect(render(<Text fontStyle='italic'>Test italic</Text>).container).toMatchSnapshot()
  expect(render(<Text textDecoration='underline'>Test underline</Text>).container).toMatchSnapshot()
})

it('should accept "variant" props', () => {
  expect(render(<Text variant='main'>Main text</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='secondary'>Secondary text</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='disabled'>Disabled text</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='link'>Link</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='h1'>Heading 1</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='h2'>Heading 2</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='h3'>Heading 3</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='h4'>Heading 4</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='h5'>Heading 5</Text>).container).toMatchSnapshot()
  expect(render(<Text variant='h6'>Heading 6</Text>).container).toMatchSnapshot()
})

it('should accept "component" props', () => {
  expect(render(<Text component='p'>This is a paragraph</Text>).container).toMatchSnapshot()
})

it('should accept style prop', () => {
  expect(render(<Text style={{ color: 'red' }}>This is red</Text>).container).toMatchSnapshot()
})

it('should accept HTMLElement props', () => {
  const { container } = render(<Text id='test-1'>Text</Text>)
  expect(container.querySelector('span').getAttribute('id')).toEqual('test-1')
})

it('should accept type argument and infer props', () => {
  render(
    <Text<AnchorHTMLAttributes<HTMLAnchorElement>> component='a' href='/'>
      Text as anchor tag
    </Text>
  )
})
