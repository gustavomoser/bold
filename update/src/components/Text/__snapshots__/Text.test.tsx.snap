// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "component" props 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
}

<div>
  <p
    class="emotion-0"
  >
    This is a paragraph
  </p>
</div>
`;

exports[`should accept "variant" props 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
}

<div>
  <span
    class="emotion-0"
  >
    Main text
  </span>
</div>
`;

exports[`should accept "variant" props 2`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #696979;
}

<div>
  <span
    class="emotion-0"
  >
    Secondary text
  </span>
</div>
`;

exports[`should accept "variant" props 3`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #AAAAB9;
}

<div>
  <span
    class="emotion-0"
  >
    Disabled text
  </span>
</div>
`;

exports[`should accept "variant" props 4`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #0069D0;
  font-weight: bold;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <span
    class="emotion-0"
  >
    Link
  </span>
</div>
`;

exports[`should accept "variant" props 5`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1.5rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <span
    class="emotion-0"
  >
    Heading 1
  </span>
</div>
`;

exports[`should accept "variant" props 6`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1.25rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <span
    class="emotion-0"
  >
    Heading 2
  </span>
</div>
`;

exports[`should accept "variant" props 7`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <span
    class="emotion-0"
  >
    Heading 3
  </span>
</div>
`;

exports[`should accept "variant" props 8`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.875rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <span
    class="emotion-0"
  >
    Heading 4
  </span>
</div>
`;

exports[`should accept "variant" props 9`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <span
    class="emotion-0"
  >
    Heading 5
  </span>
</div>
`;

exports[`should accept "variant" props 10`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.75rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <span
    class="emotion-0"
  >
    Heading 6
  </span>
</div>
`;

exports[`should accept style prop 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  color: red;
}

<div>
  <span
    class="emotion-0"
  >
    This is red
  </span>
</div>
`;

exports[`should accept text style decoration props 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <span
    class="emotion-0"
  >
    Test weight
  </span>
</div>
`;

exports[`should accept text style decoration props 2`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  font-size: 2rem;
}

<div>
  <span
    class="emotion-0"
  >
    Test size
  </span>
</div>
`;

exports[`should accept text style decoration props 3`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  color: #0069D0;
}

<div>
  <span
    class="emotion-0"
  >
    Test color
  </span>
</div>
`;

exports[`should accept text style decoration props 4`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  font-style: italic;
}

<div>
  <span
    class="emotion-0"
  >
    Test italic
  </span>
</div>
`;

exports[`should accept text style decoration props 5`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <span
    class="emotion-0"
  >
    Test underline
  </span>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
}

<div>
  <span
    class="emotion-0"
  >
    Test 1
  </span>
</div>
`;
